cpp/include/cuspatial/experimental/detail/point_linestring_nearest_points.cuh

```diff
@@ -0,0 +1,172 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuspatial/detail/utility/device_atomics.cuh>
+#include <cuspatial/detail/utility/linestring.cuh>
+#include <cuspatial/error.hpp>
+#include <cuspatial/traits.hpp>
+#include <cuspatial/vec_2d.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/binary_search.h>
+
+#include <iterator>
+#include <limits>
+#include <memory>
+#include <type_traits>
+
+namespace cuspatial {
+namespace detail {
+
+/**
+ * @internal
+ * @brief Kernel to compute the nearest point between a multipoint and multilinestring
+ *
+ * See header only API for input parameter definitions.
+ *
+ * Each thread computes the nearest point between a pair of multipoint and multilinestring.
+ * The minimum distance between the geometries are stored in `min_distance_squared` and updated
+ * when smaller is encountered. `linestring.cuh::point_to_segment_distance_squared_nearest_point`
+ * is resorted to compute the nearest point on the segment and its distance to the test point.
+ */
+template <class Vec2dItA,
+          class Vec2dItB,
+          class OffsetIteratorA,
+          class OffsetIteratorB,
+          class OffsetIteratorC,
+          class OutputIt>
+void __global__
+pairwise_point_linestring_nearest_points_kernel(OffsetIteratorA points_geometry_offsets_first,
+                                                OffsetIteratorA points_geometry_offsets_last,
+                                                Vec2dItA points_first,
+                                                Vec2dItA points_last,
+                                                OffsetIteratorB linestring_geometry_offsets_first,
+                                                OffsetIteratorB linestring_geometry_offsets_last,
+                                                OffsetIteratorC linestring_part_offsets_first,
+                                                OffsetIteratorC linestring_part_offsets_last,
+                                                Vec2dItB linestring_points_first,
+                                                Vec2dItB linestring_points_last,
+                                                OutputIt output_first)
+{
+  using T        = iterator_vec_base_type<Vec2dItA>;
+  using SizeType = iterator_value_type<OffsetIteratorA>;
+
+  auto num_pairs =
+    thrust::distance(points_geometry_offsets_first, points_geometry_offsets_last) - 1;
+  auto num_linestring_points = thrust::distance(linestring_points_first, linestring_points_last);
+
+  for (auto idx = threadIdx.x + blockIdx.x * blockDim.x; idx < num_pairs;
```

Comment:
Seeing this quadruple nested loop, I see now how important it will be to add automatic geometry iterators that know how to do the nested indexing. You want to be able to just iterate over the pairs, and for each pair, just have a for_each over the points and a nested for_each over the points of the multilinestring.

Reply:
Yes, abstracting different level of iteration is direly needed. Perhaps we can continue the discussion in #677 ?